cmd/noms/diff/diff.go

```diff
@@ -5,134 +5,146 @@
+//  * a Value is Added or Removed from a node in the graph
+//  * the type of a Value has changed in the graph
+//  * a primitive (i.e. Bool, Number, String, Ref or Blob) Value has changed
+//
+// A Difference is not returned when a non-primitive value has been modified. For
+// example, a struct field has been changed from one Value of type Employee to
+// another. Those modifications are accounted for by the Differences described
+// above at a lower point in the graph.
+//
+// If |leftRight| is true then the left-right diff is used for ordered sequences
+// - see Diff vs DiffLeftRight in Set and Map.
+//
+
+func (d Differ) Diff(v1, v2 types.Value) {
+	if !v1.Equals(v2) {
+		if !shouldDescend(v1, v2) {
+			d.sendDiff(Difference{Path: nil, ChangeType: types.DiffChangeModified, OldValue: v1, NewValue: v2})
+		} else {
+			d.diff(nil, v1, v2)
+		}
 	}
-
-	return diff(dw, types.Path{}, nil, v1, v2)
 }
 
-func diff(w diffWriter, p types.Path, key, v1, v2 types.Value) error {
+func (d Differ) diff(p types.Path, v1, v2 types.Value) bool {
 	switch v1.Type().Kind() {
 	case types.ListKind:
-		return diffLists(w, p, v1.(types.List), v2.(types.List))
+		return d.diffLists(p, v1.(types.List), v2.(types.List))
 	case types.MapKind:
-		return diffMaps(w, p, v1.(types.Map), v2.(types.Map))
+		return d.diffMaps(p, v1.(types.Map), v2.(types.Map))
 	case types.SetKind:
-		return diffSets(w, p, v1.(types.Set), v2.(types.Set))
+		return d.diffSets(p, v1.(types.Set), v2.(types.Set))
 	case types.StructKind:
-		return diffStructs(w, p, v1.(types.Struct), v2.(types.Struct))
+		return d.diffStructs(p, v1.(types.Struct), v2.(types.Struct))
 	default:
 		panic("Unrecognized type in diff function")
 	}
 }
 
-func diffLists(w diffWriter, p types.Path, v1, v2 types.List) (err error) {
+func (d Differ) diffLists(p types.Path, v1, v2 types.List) (cont bool) {
 	spliceChan := make(chan types.Splice)
-	stopChan := make(chan struct{}, 1) // buffer size of 1, so this won't block if diff already finished
+	stopChan := make(chan struct{}, 1) // buffer size of 1s, so this won't block if diff already finished
+
+	defer func() {
+		if !cont {
+			stopChan <- struct{}{}
+			// Wait for diff to stop.
+			for range spliceChan {
+			}
+		}
+	}()
 
 	go func() {
 		v2.Diff(v1, spliceChan, stopChan)
 		close(spliceChan)
 	}()
 
-	wroteHdr := false
-
 	for splice := range spliceChan {
-		if err != nil {
-			break
-		}
-
 		if splice.SpRemoved == splice.SpAdded {
 			// Heuristic: list only has modifications.
-			for i := uint64(0); i < splice.SpRemoved && err == nil; i++ {
+			for i := uint64(0); i < splice.SpRemoved; i++ {
 				lastEl := v1.Get(splice.SpAt + i)
 				newEl := v2.Get(splice.SpFrom + i)
 				if shouldDescend(lastEl, newEl) {
 					idx := types.Number(splice.SpAt + i)
-					writeFooter(w, &wroteHdr)
-					err = diff(w, append(p, types.NewIndexPath(idx)), idx, lastEl, newEl)
+					if !d.diff(append(p, types.NewIndexPath(idx)), lastEl, newEl) {
```

Comment:
I would prefer you didn't make cosmetic changes like this. Transforming the old code into the new should be:
```
cont = d.diff(append(p, types.NewIndexPath(idx)), lastEl, newEl)
```
and keep the `err != nil` check. It would make it easier to review this patch if you changed it back.